File: internal/tui/simple.go
package tui

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/charmbracelet/lipgloss"

	"github.com/aungmyatmoe/zencode/internal/config"
	"github.com/aungmyatmoe/zencode/internal/ui"
	"github.com/aungmyatmoe/zencode/internal/util"
)

// SimpleREPL is the mobile / Termux-native "open" interface.
// It prints everything to normal terminal scrollback (finger scroll + copy work great in Termux).
// Favors slash commands over hard modifier keys. Very forgiving on small screens and soft keyboards.
type SimpleREPL struct {
	cfg   *config.Config
	yolo  bool
	mode  string // Normal | Plan | YOLO
	scan  *bufio.Scanner
	theme *lipgloss.Renderer // we just use the styles directly
}

func NewSimple(cfg *config.Config) *SimpleREPL {
	yolo := cfg.Yolo
	mode := "Normal"
	if yolo {
		mode = "YOLO"
	}
	return &SimpleREPL{
		cfg:  cfg,
		yolo: yolo,
		mode: mode,
		scan: bufio.NewScanner(os.Stdin),
	}
}

func (s *SimpleREPL) Run() error {
	s.printWelcome()

	for {
		fmt.Print(ui.PromptStyle.Render("z> ") + " ")
		if !s.scan.Scan() {
			break
		}
		line := strings.TrimSpace(s.scan.Text())
		if line == "" {
			continue
		}

		// Slash commands first (easiest on mobile soft keyboard)
		if strings.HasPrefix(line, "/") {
			if s.handleSlash(line) {
				continue
			}
			// unknown command falls through or handled
		}

		s.printUser(line)

		// Simulate the agent (real loop + tools will replace this)
		s.simulateAgent(line)

		// Small pause so output doesn't feel instant on real models later
		time.Sleep(80 * time.Millisecond)
	}

	fmt.Println(ui.Dim.Render("\nbye"))
	return nil
}

func (s *SimpleREPL) printWelcome() {
	termux := isTermux()
	kind := "terminal"
	if termux {
		kind = "Termux"
	}

	fmt.Println(ui.Accent.Render("zencode") + ui.Dim.Render("  mobile-first coding agent for "+kind))
	fmt.Printf("model: %s   mode: %s\n", s.cfg.Model, s.mode)
	if s.cfg.ConfigPath != "" {
		fmt.Printf(ui.Dim.Render("config: %s\n"), s.cfg.ConfigPath)
	}

	fmt.Println(ui.Dim.Render("Type /help for commands. Everything stays in your normal scrollback (great for phones)."))
	fmt.Println()
}

func (s *SimpleREPL) printUser(text string) {
	fmt.Println(ui.UserHeader.Render("You:") + " " + text)
}

func (s *SimpleREPL) simulateAgent(userMsg string) {
	lower := strings.ToLower(userMsg)

	// Fake "thinking" line
	fmt.Print(ui.Thinking.Render("Agent thinking..."))
	time.Sleep(90 * time.Millisecond)
	fmt.Print("\r" + strings.Repeat(" ", 30) + "\r") // clear the line

	if strings.Contains(lower, "todo") || strings.Contains(lower, "plan") {
		fmt.Println(ui.Success.Render("Agent:") + " I'll track this with todos (visible and persistent in real version).")
		fmt.Println(ui.Dim.Render("  [todo_write]"))
		fmt.Println("  - [ ] Understand the request")
		fmt.Println("  - [x] Show nice mobile output")
		fmt.Println("  - [ ] Implement real search_replace + run cmd")
		fmt.Println()
		return
	}

	if strings.Contains(lower, "edit") || strings.Contains(lower, "fix") || strings.Contains(lower, "change") {
		fmt.Println(ui.Success.Render("Agent:") + " Using the precise search_replace tool (must match exactly once).")
		fmt.Println(ui.Dim.Render("  (this is the important safety property — ambiguous replaces are rejected)"))

		// Actually exercise the real safety util we have
		demoContent := "package main\n\nfunc main() {\n\tfmt.Println(\"hello\")\n}\n"
		oldStr := "\tfmt.Println(\"hello\")"
		newStr := "\tfmt.Println(\"hello from zencode on Termux\")\n\t// TODO: more useful code"

		_, res, err := util.ApplySearchReplace(demoContent, oldStr, newStr)
		if err != nil {
			fmt.Println("  (demo replace would have been rejected: " + err.Error() + ")")
		} else {
			fmt.Println(ui.DiffHeader.Render("```diff"))
			fmt.Print(res.Diff)
			fmt.Println("```")
		}
		fmt.Println()

		if !s.yolo {
			fmt.Print("Apply this change? [y/N/a(lways)] ")
			if s.scan.Scan() {
				ans := strings.ToLower(strings.TrimSpace(s.scan.Text()))
				if ans == "y" || ans == "a" {
					fmt.Println(ui.Success.Render("  → change applied (in real run we would write the file)"))
					if ans == "a" {
						s.yolo = true
						s.mode = "YOLO"
						fmt.Println(ui.Warning.Render("  YOLO mode enabled for this session"))
					}
				} else {
					fmt.Println(ui.Dim.Render("  → skipped"))
				}
			}
		} else {
			fmt.Println(ui.Success.Render("  (YOLO: would auto-apply)"))
		}
		return
	}

	// Default response
	fmt.Println(ui.Success.Render("Agent:") + " Got it. (This is the lightweight open terminal interface — perfect for Termux on phones.)")
	fmt.Println(ui.Dim.Render("  Real model calls + full tool set (read/grep/edit/shell/todo/subagent) coming next."))
	fmt.Println(ui.Dim.Render("  Try messages with 'todo', 'plan', or 'edit'. Use /help for controls."))
	fmt.Println()
}

func (s *SimpleREPL) handleSlash(line string) bool {
	cmd := strings.ToLower(strings.TrimSpace(line))
	switch {
	case cmd == "/help" || cmd == "/h" || cmd == "/?":
		s.printHelp()
		return true

	case cmd == "/yolo" || cmd == "/always-approve":
		s.yolo = !s.yolo
		if s.yolo {
			s.mode = "YOLO"
			fmt.Println(ui.Warning.Render("YOLO mode ON — edits and commands will auto-approve"))
		} else {
			s.mode = "Normal"
			fmt.Println("YOLO mode OFF")
		}
		return true

	case cmd == "/plan":
		fmt.Println(ui.Accent.Render("Plan mode") + " (stub): the agent will explore/read but not edit until you approve a plan.")
		fmt.Println("In the full version this writes a reviewable plan.md and shows it for approval.")
		return true

	case cmd == "/model" || strings.HasPrefix(cmd, "/model "):
		parts := strings.SplitN(line, " ", 2)
		if len(parts) == 2 && strings.TrimSpace(parts[1]) != "" {
			s.cfg.Model = strings.TrimSpace(parts[1])
			fmt.Printf("Model switched to %s\n", s.cfg.Model)
		} else {
			fmt.Printf("Current model: %s\n", s.cfg.Model)
			fmt.Println("Usage: /model grok-3   or   /model claude-3-5-sonnet")
		}
		return true

	case cmd == "/exit" || cmd == "/quit" || cmd == "/q":
		fmt.Println("Exiting...")
		os.Exit(0)
		return true

	case cmd == "/status":
		fmt.Printf("model=%s  mode=%s  yolo=%v\n", s.cfg.Model, s.mode, s.yolo)
		return true

	default:
		fmt.Println(ui.Dim.Render("Unknown command. Try /help"))
		return true
	}
}

func (s *SimpleREPL) printHelp() {
	fmt.Println(ui.Accent.Render("Zencode commands (mobile friendly):"))
	fmt.Println("  /help, /h          this help")
	fmt.Println("  /yolo              toggle auto-approve (dangerous but fast on phone)")
	fmt.Println("  /plan              enter plan mode (design before code)")
	fmt.Println("  /model <name>      switch model (e.g. grok-3, gpt-4o)")
	fmt.Println("  /status            show current settings")
	fmt.Println("  /exit, /quit       leave")
	fmt.Println()
	fmt.Println(ui.Dim.Render("Just type normally to talk to the agent."))
	fmt.Println(ui.Dim.Render("On Termux: volume-down often acts as Ctrl. Slash commands are easiest."))
	fmt.Println()
}

func isTermux() bool {
	// Common Termux indicators
	if os.Getenv("TERMUX_VERSION") != "" {
		return true
	}
	prefix := os.Getenv("PREFIX")
	if strings.Contains(prefix, "com.termux") {
		return true
	}
	return false
}

File: src/tui/simple.ts
import chalk from "chalk";
import Enquirer from "enquirer";
import type { Config } from "../config.js";
import { Agent } from "../agent/Agent.js";

export class SimpleREPL {
  private agent: Agent;
  private mode: "Normal" | "Plan" | "YOLO";
  private yolo: boolean = false;
  private model: string;

  constructor(private cfg: Config) {
    this.yolo = !!cfg.yolo;
    this.mode = this.yolo ? "YOLO" : "Normal";
    this.model = cfg.model;
    this.agent = new Agent(cfg);
  }

  async run(): Promise<void> {
    const isTerm = isTermux();
    const platform = isTerm ? "Termux (mobile)" : "terminal";

    console.log(
      chalk.hex("#7c3aed").bold("zencode") +
        chalk.gray(`  mobile-first coding agent for ${platform}`)
    );
    console.log(`model: ${this.model}   mode: ${this.mode}`);
    if ((this.cfg as any).configPath) {
      console.log(chalk.gray(`config: ${(this.cfg as any).configPath}`));
    }

    // Pre-flight warning for the common case (global config + Cloudflare Kimi)
    const c = this.cfg as any;
    const needsCf = this.model.startsWith('@cf/') || c.provider === 'cloudflare' || this.model.includes('kimi');
    if (needsCf && !c.cloudflareAccountId) {
      console.log(chalk.yellow('⚠  No Cloudflare credentials found for the default Kimi route.'));
      console.log(chalk.gray('   The reported config file was read but did not supply accountId/apiKey, and no matching env vars were present.'));
      console.log(chalk.gray('   Fix options:'));
      console.log(chalk.gray('     1. Add to your ~/.zshrc (or ~/.zprofile) and restart the terminal:'));
      console.log(chalk.gray('          export CLOUDFLARE_ACCOUNT_ID=6838bf50a0d8548d5945008dc7b6797c'));
      console.log(chalk.gray('          export CLOUDFLARE_API_TOKEN=cfat_...your-full-token'));
      console.log(chalk.gray('     2. Edit the config file above and ensure it contains:'));
      console.log(chalk.gray('          "provider": { "cloudflare": { "options": { "accountId": "...", "apiKey": "..." } } }'));
      console.log(chalk.gray('   Env vars (from .zshrc etc.) always win over the json file.'));
      console.log();
    }

    console.log(
      chalk.gray(
        "Everything stays in normal scrollback (finger scroll + copy work great). Use /help."
      )
    );
    console.log();

    const enquirer = new (Enquirer as any)();

    while (true) {
      let answer: any;
      try {
        answer = await enquirer.prompt({
          type: "input",
          name: "cmd",
          message: chalk.hex("#7c3aed")("z>"),
        });
      } catch {
        console.log(chalk.gray("bye"));
        break;
      }

      const trimmed = (answer.cmd || "").trim();
      if (!trimmed) continue;

      if (trimmed === "/exit" || trimmed === "/quit") {
        console.log(chalk.gray("bye"));
        break;
      }

      if (trimmed.startsWith("/")) {
        await this.handleSlash(trimmed);
        continue;
      }

      this.printUser(trimmed);

      try {
        const final = await this.agent.process(trimmed);
        console.log(chalk.green("Agent:") + " " + final);

        const todos = this.agent.getTodos ? this.agent.getTodos() : [];
        if (todos.length > 0) {
          console.log(chalk.yellow("Todos:"));
          todos.forEach((t: any) => {
            const icon = t.status === 'done' ? '✓' : t.status === 'in_progress' ? '→' : '○';
            console.log(`  ${icon} ${t.content}`);
          });
        }
      } catch (err: any) {
        const msg = (err?.message || String(err)).trim();
        console.log(chalk.red("Agent error: ") + msg.split('\n')[0]);
        // If the error already contains the Tip from the LLM layer, don't duplicate.
        if (!/Tip:/.test(msg)) {
          console.log(chalk.gray("   Check the config path above, your ~/.zshrc exports, or use /status."));
        }
      }
    }
  }

  private printUser(text: string) {
    console.log(chalk.hex("#3b82f6").bold("You:") + " " + text);
  }



  private async handleSlash(cmd: string): Promise<boolean> {
    const c = cmd.toLowerCase();

    if (c === "/help" || c === "/h" || c === "/?") {
      console.log(chalk.hex("#7c3aed")("Zencode commands (mobile-friendly):"));
      console.log("  /help            show this");
      console.log("  /yolo            toggle auto-approve (use carefully)");
      console.log("  /plan            enter plan mode (think before code)");
      console.log("  /model <name>    switch model (e.g. grok-3)");
      console.log("  /status          current settings");
      console.log("  /exit, /quit     leave");
      console.log();
      console.log(chalk.gray("Just type to chat with the agent."));
      return true;
    }

    if (c === "/yolo") {
      this.yolo = !this.yolo;
      this.mode = this.yolo ? "YOLO" : "Normal";
      console.log(this.yolo ? chalk.yellow("YOLO mode ON") : "YOLO mode OFF");
      return true;
    }

    if (c === "/plan") {
      console.log(chalk.hex("#7c3aed")("Plan mode (stub):") + " agent will only explore. Real version will produce a reviewable plan.");
      return true;
    }

    if (c.startsWith("/model")) {
      const parts = cmd.split(" ");
      if (parts[1]) {
        this.model = parts.slice(1).join(" ");
        console.log(`Model → ${this.model}`);
      } else {
        console.log(`Current model: ${this.model}`);
        console.log("Example: /model grok-3");
      }
      return true;
    }

    if (c === "/status") {
      const c: any = this.cfg;
      const hasCf = !!c.cloudflareAccountId;
      const hasKey = !!(c.apiKey && c.apiKey.length > 4);
      console.log(`model=${this.model}  mode=${this.mode}  yolo=${this.yolo}`);
      console.log(`config=${c.configPath || '(none)'}  cfCreds=${hasCf ? 'yes' : 'NO'}  apiKey=${hasKey ? 'set' : 'missing'}`);
      if (!hasCf && (this.model.startsWith('@cf/') || this.model.includes('kimi'))) {
        console.log(chalk.yellow('  → Cloudflare creds missing. See the warning above or put them in the json / env.'));
      }
      return true;
    }

    if (c === "/exit" || c === "/quit" || c === "/q") {
      console.log("Exiting...");
      process.exit(0);
    }

    console.log(chalk.gray("Unknown command. /help for list."));
    return true;
  }
}

function isTermux(): boolean {
  return !!process.env.TERMUX_VERSION || (process.env.PREFIX || "").includes("com.termux");
}
